docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,408 @@
+from this API and not just the north south traffic.
+
+### API Details
+
+* A new API `NetworkQoS` under the `k8s.ovn.org/v1` version will be added to `go-controller/pkg/crd/networkqos/v1`.
+This would be a namespace-scoped CRD:
+
+```go
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:resource:path=networkqoses
+// +kubebuilder::singular=networkqos
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// NetworkQoS is a CRD that allows the user to define a DSCP marking and metering
+// for pods ingress/egress traffic on its namespace to specified CIDRs,
+// protocol and port. Traffic belong these pods will be checked against
+// each Rule in the namespace's NetworkQoS, and if there is a match the traffic
+// is marked with relevant DSCP value and enforcing specified policing/shaping
+// parameters.
+type NetworkQoS struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec   Spec   `json:"spec,omitempty"`
+    Status Status `json:"status,omitempty"`
+}
+
+// Spec defines the desired state of NetworkQoS
+type Spec struct {
+    // networkAttachmentName selects the network-attachment-definition for
+    // which the QoS Rules will be applied. The NAD can be of any type Layer-3,
+    // Layer-2 or Localnet. If not specified, the default cluster-wide network
+    // will be used.
+    // +optional
+    NetworkAttachmentName string `json:"networkAttachmentName,omitempty"
+
+    // podSelector applies the NetworkQoS rule only to the pods in the namespace whose label
+    // matches this definition. This field is optional, and in case it is not set
+    // results in the rule being applied to all pods in the namespace.
+    // +optional
+    PodSelector metav1.LabelSelector `json:"podSelector,omitempty"`
+
+    // egress a collection of Egress NetworkQoS rule objects
+    Egress []Rule `json:"egress"`
+}
+
+type Rule struct {
+    // priority The NetworkQoS rule’s priority. Rules with numerically higher
+    // priority take precedence over those with lower. If two NetworkQoS
+    // rules with the same priority both match, then the one
+    // actually applied to a packet is undefined.
+    // +kubebuilder:validation:Maximum:=32767
+    // +kubebuilder:validation:Minimum:=0
+    Priority int `json:"priority"`
+
+    // dscp marking value for matching pods' traffic.
+    // +kubebuilder:validation:Maximum:=63
+    // +kubebuilder:validation:Minimum:=0
+    DSCP int `json:"dscp"`
+
+    // classifier The classifier on which packets should match
+    // to apply the NetworkQoS Rule.
+    // This field is optional, and in case it is not set the rule is applied
+    // to all egress traffic regardless of the destination.
+    // +optional
+    Classifier Classifer `json:"classifier"`
+
+    // +optional
+    Bandwidth Bandwidth `json:"bandwidth"`
+}
+
+type Classifer struct {
+    // +optional
+    To []Destination `json:"to"`
```

Comment:
.

Reply:
what am i missing here ? I can't make out what `.` meant for :)